### PR TITLE
[TYP] Add tool for running stubtest

### PR DESCRIPTION
**Pull request by ksunden**
## PR summary

Walks the ast to find delete_parameter calls and automatically ignores them as their signature is modified at runtime

~~Could potentially be further modified to respect the alias behavior, which is the majority of the current allowlist~~ (Added this feature)

Makes running stubtest easier as now it is just 'python tools/stubtest.py', rather than a long incantation with multiple arguments

This was discussed in part in #26894

## PR checklist

- [ ] "closes #0000" is in the body of the PR description to [link the related issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [ ] new and changed code is [tested](https://matplotlib.org/devdocs/devel/testing.html)
- [ ] *Plotting related* features are demonstrated in an [example](https://matplotlib.org/devdocs/devel/document.html#write-examples-and-tutorials)
- [ ] *New Features* and *API Changes* are noted with a [directive and release note](https://matplotlib.org/devdocs/devel/coding_guide.html#new-features-and-api-changes)
- [ ] Documentation complies with [general](https://matplotlib.org/devdocs/devel/document.html#write-rest-pages) and [docstring](https://matplotlib.org/devdocs/devel/document.html#write-docstrings) guidelines

### CODE REVIEW

**Review by anntzer — APPROVED** [review]

**Comment by QuLogic** [comment]
I'm curious whether this works when matplotlib is installed normally (non-editable), as that is something that I needed to do on the Meson branch.

**Comment by ksunden** [comment]
Yes, it does appear to work on installed mpl

**Comment by ksunden** [comment]
Stubtest takes longer than I'm really willing to put in a pre-commit hook (it runs mypy first and then does its checks which involve importing)

On my machine this takes over a minute in total:

- ~3 seconds for the newly added ast parsing/generating of ignores in this script
- ~15 seconds for mypy to run (judging based on running it independently)
- ~50 seconds for stubtest itself to run

I could _maybe_ justify normal mypy as a pre-commit (though running _all_ of our precommit hooks (`precommit run --all`) currently takes less than 5 seconds (and the largest chunk is flake8, which I'm interested in replacing with faster ruff), so even 15 for mypy is pushing it a bit further, though pre-commit may run in more parallel)

Stubtest (well, at least this script as written) does not have as good of ways to only run on certain files. (you also specify by the import path because it is checking runtime behavior, so at least some manipulation would have to be done to the filepaths, though possible).

All of these numbers are on my (pretty beefy) desktop machine, and would hesitate to do something that makes it worse.

In all, not totally opposed to mypy, especially since I think it scopes to the edited files, so the worst case of 15 seconds is not likely to happen. but stubtest is probably good where it is at CI level and a tool that you can run locally if you either expect it to possibly fail or are diagnosing a failure from CI.

**Comment by ksunden** [comment]
True, but whether it is worth it is dependent on not only the speedup of the failing path, but also the slowing down of the passing path and the proportion of changes which go down each path.

I am probably a bit of an outlier in that my changes are very bimodal with really small changes (where an extra 15 seconds is frustrating) and really large changes (where I'm running these tools locally myself multiple times before I get to committing even), so perhaps not the best one to make that judgement.

Again, I think I'd support adding mypy, but not stubtest, as a happy medium (especially if it scopes to edited files by default).

File: tools/stubtest.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+                    self.output.write(
+                        "\n".join(
+                            f"{'.'.join(parts)}.[gs]et_{a}\n" for a in substitutions
+                        )
+                    )
+        for child in ast.iter_child_nodes(node):
+            self.visit(child)
+
+
+with tempfile.NamedTemporaryFile("wt", delete=False) as f:
```
Comment (QuLogic) [inline tools/stubtest.py:55]: Ensure this cleans up like the context manager:
```suggestion
# Use delete_on_close when Python 3.12 is minimum.
f = tempfile.NamedTemporaryFile("wt", delete=False)
try:
```

File: tools/stubtest.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+            "stubtest",
+            "--mypy-config-file=pyproject.toml",
+            "--allowlist=ci/mypy-stubtest-allowlist.txt",
+            f"--allowlist={f.name}",
+            "matplotlib",
+        ],
+        cwd=root,
+        env=os.environ | {"MPLBACKEND": "agg"},
+    )
+    try:
```
Comment (QuLogic) [inline tools/stubtest.py:79]:
```suggestion
    )
finally:
    f.close()
    try:
```

File: tools/stubtest.py (hunk 1 above)
Comment (anntzer) [reply tools/stubtest.py:55]: Alternatively you can just use a TemporaryDirectory instead and create a file with whatever name you want in that directory, to get both proper autodeletion without problems with close()ing the file.  We use that pattern quite a few times in the library itself.

Comment (ksunden) [reply tools/stubtest.py]: Done

File: tools/stubtest.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+                for child in ast.iter_child_nodes(node):
+                    child.parent = node
+
+            v.visit(tree)
+    proc = subprocess.run(
+        [
+            "stubtest",
+            "--mypy-config-file=pyproject.toml",
+            "--allowlist=ci/mypy-stubtest-allowlist.txt",
+            f"--allowlist={str(p)}",
```
Comment (QuLogic) [inline tools/stubtest.py:73]: Isn't this redundant?
```suggestion
            f"--allowlist={p}",
```

**Review by QuLogic — APPROVED** [review]

File: tools/stubtest.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+            f"--allowlist={p}",
+            "matplotlib",
+        ],
+        cwd=root,
+        env=os.environ | {"MPLBACKEND": "agg"},
+    )
+    try:
+        os.unlink(f.name)
+    except OSError:
+        pass
```
Comment (QuLogic) [inline tools/stubtest.py:82]: Oops, I just realized this is still here; it can probably be deleted now that you're using `TemporaryDirectory`?

**Pull request merged**